### ISSUE TITLE
Developer Experience Degradation

### ISSUE THREAD

**Issue by Skurkify** [facebook/flow#7647]
Hi Flow maintainers. Before I jump into describing my concerns about developer experience, I want to express my gratitude for all the work you've put into Flow over the years. Thanks for everything you've done. With that said, please believe that this communication is coming from a good place.

I'm concerned the Flow road map isn't prioritizing developer experience. I've purposefully delayed updating my org from flow 0.83.0 to a more recent version because the requirement for type annotations to be specified in the latest flow versions when using libraries such as reselect and react-redux, is a huge downgrade to developer experience. When using reselect and react-redux in 0.83.0, annotations aren't required. They're powered by inference, and I believe that is the experience that developers would prefer. From what I understand, the Flow team moved in this direction to capture some additional correctness. I feel the additional correctness captured was not worth the developer experience we lost.

Is there a plan to recover the lost inference and easier developer experience that was present in 0.83.0?

**Comment by PinkaminaDianePie** [comment]
Yes, `Developer Experience` is the main selling point for any tool right now, so it would be nice to have such priorities. I believe the Flow team won't return to the old way of providing types, but this thing can be solved on the tooling side - language server can just infer these annotations on demand and autosuggest them to you, so your code will still have them, but in 99% of cases you will be able to do it automatically. File level boundaries are great improvements for performance, so for me it's nice trade-off to keep both DX and performance.

**Comment by goodmind** [comment]
Maybe directory level boundaries would be possible?

**Comment by wgao19** [comment]
Hi @Skurkify I also found it an ordeal to upgrade flow past 0.85. I was personally blocked for weeks until I finally realize how one way was way easier than another and managed to clear our hundreds of implicitly instantiation errors. After I figured them out, I realize the gain in accuracy in typing is indeed valuable and the error messages help to locate the actual type errors a lot easier.

And yeah, I also notice some dev experience issue around this process. What really baffles me is a lack of guidance I guess? I couldn't find any blog post, not even a lot of questions asked about this issue. How can that be true?

**Comment by PinkaminaDianePie** [comment]
@wgao19 they published some guidance a few weeks ago: https://medium.com/flow-type/upgrading-flow-codebases-40ef8dd3ccd8

**Comment by nmote** [comment]
Here's a blog post with more info on the specific change made in 0.85, and why it was done:

https://medium.com/flow-type/asking-for-required-annotations-64d4f9c1edf8

Basically, these annotations should have always been required, but due to a bug, they were not. This led Flow to unpredictably miss legitimate type errors. We understand that this makes some things more difficult, but we believe that the improved reliability of typechecking outweighs the inconvenience of adding additional type annotations.

Developer experience is definitely important to us. In this case, we had to weigh whether it was a better developer experience for Flow to silently miss real type errors, or for Flow to require additional type annotations. We believe that silently missing errors is worse.

Also note that we are planning, at some point, to require type annotations at all module boundaries (not just input positions). This is being done mainly to improve performance, though it has other benefits as well. This is still a long way out, though, and we'll make sure to have tools available to help people make the transition well before we start requiring these annotations.

**Comment by Skurkify** [comment]
Thanks for the responses, all.
